src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -111,25 +108,22 @@ public bool TryQueueNewStream(SafeMsQuicStreamHandle streamHandle, QUIC_STREAM_O
 
             public bool TryAddStream(MsQuicStream stream)
             {
-                lock (this)
+                // Refuse new streams if Dispose was called.
+                if (Volatile.Read(ref _closing))
                 {
-                    if (_closing)
-                    {
-                        return false;
-                    }
-
-                    StreamCount++;
-                    return true;
+                    return false;
                 }
+
+                return Interlocked.Increment(ref StreamCount) > 0;
```

Comment:
Is `> 0` here for catching the race when `_closing` was set to true after we checked it on line 112? Shouldn't we revert `Increment` if we found out it was negative? Also, the logic with additional decrement seems fragile to me... I can imagine a situation here when two streams wanted to do that at the same time, and if it was already `-1`, only the first stream will catch `> 0` check, and the second stream may get a false-positive -- even if we will revert increment of the first stream, because reverting wouldn't be atomic. Or am I missing something?

Reply:
Seems like we should be using an Interlocked.CompareExchange loop instead?

Reply:
Can you be more specific on the suggestion @stephentoub? 
The construct is probably not great and 
```c#
Interlocked.Increment(ref StreamCount);
return true;
```
 would be better. 


The important part is to not free connection when there are pending streams. So we count streams and if `Dispose` is called when there are active streams, we defer cleanup until last stream is gone. 
So the `RemoveStream` needs to know when least stream is removed if cleanup should happen or not. 

I was originally thinking about removing the `_closing` and relying only on single counter. 
The check for `_closing` in AddStream is most likely opportunistic to not accept more work when Dispose is called but I don't think it matters for MsQuic. Since the dispose calls abort, the streams would be cleaned later with error anyway.